[DEV-130] Add comment replies

libre-man opened the pull request:
This PR makes it possible to reply to inline feedback comments. This PR also changes the format of (new) inline comments to markdown, and adds user settings.

### TODO
- [x] Add new backend tests
- [ ] Create test stories for the manual user tests.

DEV-130 #done

olmokramer commented on cypress/integration/tests/sidebar.spec.js:
```diff
@@ -66,20 +66,20 @@ context('Sidebar', () => {
             cy.visit('/');
             openProfile();
-            cy.get('input[name=full-name]').clear()
+            cy.get('input[name=full-name]').clear().setText('');
```
`setText` already does a `clear`, right? Is the `clear` also needed for cypress to behave?

libre-man replied:
No it isn't I just added the `setText` because cypress was not behaving and I didn't realize the existing clear could be removed.

libre-man commented on cg_sqlalchemy_helpers/types.py:
```diff
@@ -691,7 +735,8 @@ def scalar(self) -> bool:
 _MyQuery = MyQuery
 
-if t.TYPE_CHECKING:
+MYPY = False
```
No because when building docs `t.TYPE_CHECKING` is also `True`, but because of that sqlalchemy fails (as expected), but using the `getattr` trick as elsewhere in the code doesn't work because we can't say to mypy that it will always evaluate to `True`, which we can do using this `MYPY` trick.

libre-man commented on migrations/versions/.#3027bc9b0708_.py:
```diff
@@ -0,0 +1 @@
+thomas@Libertus.17086:1585139225
```
Nope will remove.

libre-man commented on psef/auth.py:
```diff
@@ -805,25 +904,185 @@ def ensure_can_edit_members_of_group(
+        ):
+            return
+
+        if self._is_own_reply:
```
Maybe in the future when we make it possible to remove users from a course.

libre-man replied:
I just wanted to add it to be complete, but we can of course remove it.

olmokramer replied:
Fair enough. No let's keep in just in case, I was mostly curious.

libre-man commented on psef/mail_templates/notification.j2:
```diff
@@ -0,0 +1,33 @@
+{% endblock %}
+
+{% block footer %}
+    You received this email because {{ reasons }}. You can edit your
```
Good idea.

libre-man commented on .github/scripts/run_backend_tests.bash:
```diff
@@ -3,6 +3,7 @@ set -o xtrace
 export PYTHONPATH="$PWD"
 export GITHUB_ACTIONS=true
 sudo chown -R "$USER":"$(id -gn)" /tmp/
+printf '[Back-end]\nexternal_url = http://localhost:1234' > config.ini
```
Agree.

libre-man commented on migrations/versions/07eb464c56886d188ca8_.py:
```diff
@@ -0,0 +1,29 @@
+"""""Creating "can_add_own_inline_comments" permission.
```
Lets discuss that offline (or really online in a call) for a bit, as this is a bit tricky. Anyway: if we don't give existing users this permission TA's will still be able to grade.

libre-man replied:
Ok we did this in a call, and decided that the current way is fine.

olmokramer commented on psef/models/user_setting.py:
```diff
@@ -0,0 +1,339 @@
+
+logger = structlog.get_logger()
+
+MYPY = False
```
Should this be removed? There's a check later on `if MYPY: ...`

libre-man replied:
See comment above

olmokramer commented on psef/models/user_setting.py:
```diff
@@ -0,0 +1,339 @@
+
+
+EnabledEmailNotificationTypes = Literal[EmailNotificationTypes.  # pylint: disable=invalid-name
+                                        direct, EmailNotificationTypes.
```
Thanks, yapf!

olmokramer commented on psef/tasks.py:
```diff
@@ -713,6 +725,107 @@ def _add_1(first: int, second: int) -> int:  # pragma: no cover
+            except Exception:  # pragma: no cover
+                # This happens if mail sending fails or if the user has no
+                # e-mail address.
+                # TODO: make this exception more specific
```
Should this be done now or can it wait?

libre-man replied:
No this can wait, we have the same issue already at more places when sending email. But maybe we should simply drop this crappy library for sending email anyway.

olmokramer commented on psef/v1/code.py:
```diff
@@ -219,7 +225,7 @@ def get_file_url(file: models.FileMixin[object]) -> str:
 
 def get_feedback(file: models.File, linter: bool = False) -> _FeedbackMapping:
-    """Returns the :class:`.models.Comment` objects attached to the given
+    """Returns the :class:`.models.CommentBAse` objects attached to the given
```
s/CommentBAse/CommentBase/

olmokramer commented on psef/v1/submissions.py:
```diff
@@ -253,80 +292,134 @@ def delete_submission(submission_id: int) -> EmptyResponse:
-                authors[comment.file_id][comment.line] = comment.user
+    :query boolean with_replies: If considered true (see
+        :func:`.helpers.request_arg_true`) feedback is send including
+        replies. Please note that passing this as true is deprecated.
```
Shouldn't that be "passing this as _false_ is deprecated"?

libre-man replied:
Yep haha

olmokramer commented on src/components/CgLogo.vue:
```diff
@@ -1,12 +1,9 @@
-         :src="logoSrc"
-         alt="CodeGrade" />
-</template>
+<script lang="ts">
```
Now that this is a script-only component, we can get rid of the `<script>` tags if we rename it `src/components/CgLogo.ts` right?

olmokramer commented on src/components/FeedbackArea.vue:
```diff
@@ -1,862 +1,231 @@
+    <div @click.prevent="addReply"
+         class="add-reply text-muted p-2 border rounded mt-2 border-top"
+         v-if="showReply">
+        Reply&hellip;
```
Maybe "Click to reply..."?

libre-man replied:
Good idea

olmokramer commented on src/components/FeedbackArea.vue:
```diff
@@ -1,862 +1,231 @@
-        }
+<script lang="ts">
+    import { Vue, Component, Prop, Watch } from 'vue-property-decorator';
+import { Getter, Action } from 'vuex-class';
```
The indentation on either this or the previous line is wrong

libre-man replied:
Thanks, the previous one, my emacs config does this for some reason for ts lang scripts.

olmokramer commented on src/components/FeedbackArea.vue:
```diff
@@ -1,862 +1,231 @@
-        deleteFeedbackError() {
-            this.feedbackDisabled = false;
-        },
+    hiddenReplies: Set<number> = new Set(this.nonDeletedReplies.slice(1, -1).map(r => r.trackingId));
```
Should this not be a getter, since `this.nonDeletedReplies` is also a getter? Or am I missing something?

libre-man replied:
No, this should not be a getter and I should probably add a comment explaining why. The gist of it is this: lets say we have three replies, in this case the middle one would be hidden, if I now click reply to reply to the last reply I want to see that reply while writing my new one. If this would be a getter the new reply would become hidden as soon as I start writing my new reply. We could of course add logic to check for this using the `id` prop, but then when I would save my new reply the old one would be hidden.

Personally I really like the behavior the current impl. creates: it determines what will be hidden on the first render, and it shouldn't touch it after that.

olmokramer replied:
I see, and I agree with that behaviour :+1:

olmokramer commented on psef/db_locks.py:
```diff
@@ -0,0 +1,47 @@
+"""
```
I only just realized, but this file, and maybe others, are missing the license header.

libre-man replied:
I totally agree, but this file is definitely not missing the license header, look at line 7 ;)

olmokramer replied:
I see. I thought we put those on the first line in each file as a comment

olmokramer commented on src/components/FeedbackReply.vue:
```diff
@@ -0,0 +1,807 @@
+</template>
+
+<script lang="ts">
+    import { Vue, Component, Prop, Ref, Watch } from 'vue-property-decorator';
```
Again the first line in the <script> is indented wrong

olmokramer commented on src/components/FeedbackReply.vue:
```diff
@@ -0,0 +1,807 @@
+        FeedbackReplyHistory,
+    },
+})
+
```
Can you remove this empty line to make it clearer that the decorator above applies to this class?

libre-man replied:
Yeah, good catch

olmokramer commented on src/components/FileViewer.vue:
```diff
@@ -181,9 +181,9 @@ export default {
-                this.fileData &&
-                this.fileContent &&
-                this.fileContent.byteLength === 0
+                    this.fileData &&
```
Yep, definitely "prettier"...

olmokramer commented on src/components/ImageViewer.vue:
```diff
@@ -137,7 +137,7 @@ export default {
     .img {
         display: inline-block;
-        max-width: 100%;
+        // max-width: 100%;
```
Please remove

olmokramer commented on src/components/NotificationSettings.vue:
```diff
@@ -0,0 +1,170 @@
+                    </li>
+                    <li>
+                        <b>daily</b>: Send a daily digest email with all new
+                        comments. If there are no new comments no email is send.
```
s/send/sent/

olmokramer commented on src/components/Sidebar/index.vue:
```diff
@@ -23,11 +23,18 @@
+                    <icon name="bell"
+                          :scale="0.75"
+                          class="notification-bell"
+                          :style="{ opacity: entry.hasNotification() ? 1 : 0 }"
```
Can't we move this check to the v-if?

libre-man replied:
No because then we wouldn't have the animation we have now.

olmokramer replied:
Right. Didn't notice it had an animation going on

olmokramer commented on src/components/UserInfo.vue:
```diff
@@ -1,7 +1,7 @@
 <template>
 <div class="userinfo">
-    <loader class="col-md-12 text-center" v-if="loading"/>
+    <loader class="col-md-12 text-center" v-if="loading" :scale="1"/>
```
I would prefer a scale of 2 but we can discuss that later, if ever

olmokramer commented on src/style.less:
```diff
@@ -274,7 +274,7 @@ body {
 .show > .dropdown-menu {
     max-height: 50vh;
-    overflow-y: scroll;
+    overflow-y: auto;
```
Oh it was us that set that ridiculous rule? Wow.

olmokramer commented:
Why was static/img/codegrade-inv.png changed?

libre-man commented:
> Why was static/img/codegrade-inv.png changed?

Because it didn't exist yet

libre-man commented on build/webpack.prod.conf.js:
```diff
@@ -23,22 +24,16 @@ var webpackConfig = merge(baseWebpackConfig, {
     path: config.build.assetsRoot,
-    filename: utils.assetsPath('js/[name].[chunkhash].js'),
-    chunkFilename: utils.assetsPath('js/[id].[chunkhash].js')
+    // filename: utils.assetsPath('js/[name].[chunkhash].js'),
```
Agree

libre-man commented on build/webpack.prod.conf.js:
```diff
@@ -102,7 +82,37 @@ var webpackConfig = merge(baseWebpackConfig, {
+    },
+    runtimeChunk: 'single',
+
+    // minimizer: [
```
Yes

olmokramer commented:
Of course. Man I was really getting tired :P

olmokramer commented:
Some things I noticed during testing.

These I think we should fix before releasing:

- The cancel button is very dangerous, as it undoes all your edits without asking for confirmation. We should either make it a submit button with a confirmation, give it class `danger-outline` (or `outline-danger`, I always forget), or do something else to indicate that you should think twice before clicking it.

- The replies need fixing in dark mode (I know this is on your list of things to do, just including it here for posterity).
![dark-mode](https://user-images.githubusercontent.com/3612514/79068090-72391800-7cc4-11ea-8ff7-86148898e1d0.png)
  - The header should not have such a dark background
  - The buttons should always have the `@color-primary` background (or maybe it's easier to set their background to `transparent !important`), and not change their background when hovered


And the following can wait until we have time to discuss them after the release:

- When adding the first reply, the card containing the textarea has more left-margin than the comment that you're replying to:
![top-level-reply](https://user-images.githubusercontent.com/3612514/79066665-ff2aa400-7cb9-11ea-9c3d-b8b26ab12055.png)

- The font size in the text area does not take the fontSize preference into account:
![huge-text](https://user-images.githubusercontent.com/3612514/79066682-30a36f80-7cba-11ea-85c1-23fc91170ba2.png)

- I think it would be nice if we use the same style of highlighting as we do for focused inputs when we highlight a comment:
![highlight](https://user-images.githubusercontent.com/3612514/79066859-54b38080-7cbb-11ea-9cdb-b5dc533ee0c4.png)

- The text "You have no snippets" should have class `text-muted font-italic` as we do in the rest of CodeGrade when some list is empty. Also, I think it would be better if the text in the dropdown headers were black instead of muted, and I think a bit of extra margin (0.25 or 0.5 rem) between the right side of the snippet (reply) button and the card would look nicer:
![snippet-dropdown](https://user-images.githubusercontent.com/3612514/79066967-1a96ae80-7cbc-11ea-9d6f-89184d17e245.png)

- It would be nice if the save/cancel/delete buttons were also available in the preview tab.

- I don't think the edits history table needs a `.table-striped`, or at least it would be nice if the first entry in the list had a white background, in case there's only a single entry. Also, the diff column takes an unnecessary amount of horizontal space, causing the time and author name to wrap, but there may not be much we can do about that because HTML tables...
![edits-popover](https://user-images.githubusercontent.com/3612514/79067032-c3450e00-7cbc-11ea-90a1-79a0746571f3.png)

- The edits popover does not go away if I click somewhere in the page; I have to click the icon again.

- The margin below the save/cancel/delete buttons could be reduced:
![button-margin](https://user-images.githubusercontent.com/3612514/79067147-bd9bf800-7cbd-11ea-962a-656d60027732.png)

- It is unfortunate that the snippet popover is cut off when the page is not high enough. Can we set the `container` of the popover to `#submission-page`? Does that fix it?
![snippet-popover-cutoff](https://user-images.githubusercontent.com/3612514/79067210-1bc8db00-7cbe-11ea-9d98-32754de11d54.png)

- Snippets are always inserted at the end of the comment. Since inserting snippets in the middle of a comment has been broken for a long time anyway, that's not a big deal. But I think it would be nice if we inserted each snippet on a new line. Then it becomes easier to reorder them, because you can now select the entire snippet text with a triple-click. You would still get the same behaviour, because markdown renders adjacent non-empty lines as a single paragraph.

- After deleting a reply the submit button suddenly gets its usual background, has lots of vertical padding, and no horizontal padding:
![delete-done](https://user-images.githubusercontent.com/3612514/79067431-7f9fd380-7cbf-11ea-80f3-6626627505aa.png)

- The snippet selector (that appears when pressing <kbd>Tab</kbd>) does not have a top border:
![snippet-selector](https://user-images.githubusercontent.com/3612514/79067529-4025b700-7cc0-11ea-9e24-c241aa0d9890.png)

- Sometimes the snippet selector  (that appears when pressing <kbd>Tab</kbd>) does not close after selecting a snippet, but I have not been able to reliably reproduce this.

- I still think we should look for another icon for the snippet dropdown, as we use the reply arrow in quite a few places, and in all of those places it means "reset this input/set of inputs", so I fear people might think it will reset the text in the text area to what it was when they started editing, and never even think about clicking it.

- Speaking of which, it would be nice to have a button to reset the text to what it was when you started editing.

- The image viewer takes up the entire height available to it:
![image-viewer-height](https://user-images.githubusercontent.com/3612514/79067797-4b79e200-7cc2-11ea-9ad7-a6a9f0f3764e.png)

- I think the "show more replies" button also looks nice as a link, which also allows us to save some more vertical space there:
![show-more-replies](https://user-images.githubusercontent.com/3612514/79067894-2df94800-7cc3-11ea-999b-d60de057ff65.png)

- I think it might look nice to use the "eye" icon in the "mark as read" button.

- Clicking the delete button above a reply that is not being edited puts the confirmation popover in a very strange posotion, covering the delete button itself:
![delete-popover](https://user-images.githubusercontent.com/3612514/79068005-da3b2e80-7cc3-11ea-945a-175bf350a04e.png)

- The "comment edited" indicator (the asterisk) is very small because a `<sup>` has CSS `font-size: 75%`, making it very hard to exactly get my mouse on top of it to get the time tooltip. I think it still looks good (better, actually) with `font-size: 100%`. And maybe we should also add some horizontal padding to it, maybe combined with some negative margin, to increase the hoverable area.
![edit-star](https://user-images.githubusercontent.com/3612514/79068211-8a5d6700-7cc5-11ea-9c53-d3dd5d17700c.png)

libre-man commented:
> Some things I noticed during testing.
> 
> These I think we should fix before releasing:
> 
> The cancel button is very dangerous, as it undoes all your edits without asking for confirmation. We should either make it a submit button with a confirmation, give it class `danger-outline` (or `outline-danger`, I always forget), or do something else to indicate that you should think twice before clicking it.

Fixed

> The replies need fixing in dark mode (I know this is on your list of things to do, just including it here for posterity).
>       
> The header should not have such a dark background
> The buttons should always have the `@color-primary` background (or maybe it's easier to set their background to `transparent !important`), and not change their background when hovered

Fixed 

> And the following can wait until we have time to discuss them after the release:
> 
>  When adding the first reply, the card containing the textarea has more left-margin than the comment that you're replying to:

This is kind on purpose, as the card aligns with all replies (which are slightly indented. We could remove this indentation for the first reply, but I kinda like the current look. I should come back to this later.

> The font size in the text area does not take the fontSize preference into account:
Will fix later.

>  think it would be nice if we use the same style of highlighting as we do for focused inputs when we highlight a comment:

Personally I like this highlighting, so we will discuss this later.

> The text "You have no snippets" should have class `text-muted font-italic` as we do in the rest of CodeGrade when some list is empty. Also, I think it would be better if the text in the dropdown headers were black instead of muted, and I think a bit of extra margin (0.25 or 0.5 rem) between the right side of the snippet (reply) button and the card would look nicer:

I have fixed the classes, and the empty list when you filtered out all the snippets. The text of the headers is standard bootstrap vue, so I didn't change that. I have also not done anything with the margins for now.

> It would be nice if the save/cancel/delete buttons were also available in the preview tab.

I've moved them to their container that is always visible: 
![image](https://user-images.githubusercontent.com/12597247/79070740-095b9b00-7cd8-11ea-8c68-eec8087ed867.png)


> I don't think the edits history table needs a `.table-striped`, or at least it would be nice if the first entry in the list had a white background, in case there's only a single entry. Also, the diff column takes an unnecessary amount of horizontal space, causing the time and author name to wrap, but there may not be much we can do about that because HTML tables...

Agree about the striped, removed that. I have no idea how to improve the styling furhter, maybe the entire thing shouldn't even be in a popover.
 
> The edits popover does not go away if I click somewhere in the page; I have to click the icon again.

I forgot to add `blur` again after debugging, is fixed :)

> The margin below the save/cancel/delete buttons could be reduced:

Fixed with the an earlier point.

> 
> It is unfortunate that the snippet popover is cut off when the page is not high enough. Can we set the `container` of the popover to `#submission-page`? Does that fix it?

Yeah, if that thing where a popover, which it is not. We should simply rewrite the component sometime soon, but that should be easier now that it is isolated.

> Snippets are always inserted at the end of the comment. Since inserting snippets in the middle of a comment has been broken for a long time anyway, that's not a big deal. But I think it would be nice if we inserted each snippet on a new line. Then it becomes easier to reorder them, because you can now select the entire snippet text with a triple-click. You would still get the same behaviour, because markdown renders adjacent non-empty lines as a single paragraph.

Luckily that was an easy fix, so they are now inserted on a newline, except when the existing text ends with a whitespace char, in that case it is simply appended.

> After deleting a reply the submit button suddenly gets its usual background, has lots of vertical padding, and no horizontal padding:

Yeah fixed that, kinda. It still looks weird but far less.

> The snippet selector (that appears when pressing Tab) does not have a top border:

Will fix later.

> Sometimes the snippet selector  (that appears when pressing Tab) does not close after selecting a snippet, but I have not been able to reliably reproduce this.

Will fix later.
 
> I still think we should look for another icon for the snippet dropdown, as we use the reply arrow in quite a few places, and in all of those places it means "reset this input/set of inputs", so I fear people might think it will reset the text in the text area to what it was when they started editing, and never even think about clicking it.

Agree, but lets do that later!

> Speaking of which, it would be nice to have a button to reset the text to what it was when you started editing.

You can kind of do that with cancel though, so I personally wouldn't really be in favor of adding another button to do that.

> The image viewer takes up the entire height available to it.

Yeah, I have no clue how to fix that, but imo it is really the end of the world. So lets fix it later.

> I think the "show more replies" button also looks nice as a link, which also allows us to save some more vertical space there:

Agree and changed.

> I think it might look nice to use the "eye" icon in the "mark as read" button.
Agree and changed.

>  Clicking the delete button above a reply that is not being edited puts the confirmation popover in a very strange posotion, covering the delete button itself:

This is a bit better now, so lets postpone this.

> The "comment edited" indicator (the asterisk) is very small because a `<sup>` has CSS `font-size: 75%`, making it very hard to exactly get my mouse on top of it to get the time tooltip. I think it still looks good (better, actually) with `font-size: 100%`. And maybe we should also add some horizontal padding to it, maybe combined with some negative margin, to increase the hoverable area.

Agree that is too small, but I don't have a fix for now.

olmokramer commented on src/components/FeedbackReply.vue:
```diff
@@ -131,6 +116,34 @@
         </b-tabs>
+
+                    <div class="save-button-wrapper">
+                        <submit-button
```
This can now be a `<cg-submit-button>`, right?

olmokramer commented on src/components/FeedbackReply.vue:
```diff
@@ -131,6 +116,34 @@
+                                          variant="outline-primary"
+                                          label="Cancel"/>
+
+                        <submit-button :submit="submitFeedback"
```
Same as above

libre-man replied:
Fixed them all

olmokramer commented on src/components/FeedbackReply.vue:
```diff
@@ -394,6 +409,13 @@ Do you want to overwrite it?`;
     }
 
+    get cancelEditingConfirmMessage(): string {
+        if (this.internalReply.message !== this.reply.message || this.internalReply.id == null) {
```
Maybe later we may also want to check if any edits were made, and not show the confirmation if there were no edits.

libre-man replied:
Wait we already do that right? If no edits are made the `.message` properties are the same right?

olmokramer commented:
Looking good now! One more thing I noticed: the save/cancel/delete buttons now have very bright colors in dark mode:
![button-colors](https://user-images.githubusercontent.com/3612514/79072623-53e21500-7ce2-11ea-86de-7073eeec0ca2.png)

Was that intended?

libre-man commented:
> Looking good now! One more thing I noticed: the save/cancel/delete buttons now have very bright colors in dark mode:
> ![button-colors](https://user-images.githubusercontent.com/3612514/79072623-53e21500-7ce2-11ea-86de-7073eeec0ca2.png)
> 
> Was that intended?

Yeah, as this is the normal color for primary buttons in dark mode...
